File: Java/Silver/1149.java
import java.io.*;
import java.util.*;

public class Main {

    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static int[][] costMap;

    private static int[][] accCostMap;

    private static int houseCount = 0;

    public static void main(String[] args) throws IOException {

        houseCount = Integer.parseInt(br.readLine());
        costMap = new int[houseCount][3];
        accCostMap = new int[houseCount][3];

        for (int n = 0; n < houseCount; n++) {
            costMap[n] = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        }

        int minCost = calculateCost();

        System.out.println(minCost);
        br.close();
    }

    private static int calculateCost() {
        int[] calculatedCosts = new int[3];
        for (int i=0;i<3;i++) calculatedCosts[i] = calculator(0,i);
        return Arrays.stream(calculatedCosts).min().getAsInt();
    }

    private static int calculator(int y, int x) {

        if (y == houseCount) return 0;

        if (accCostMap[y][x] != 0) return accCostMap[y][x];

        int minCost = 987654321;
        for(int idx = 0; idx < 3; idx++) {
            if(idx==x) continue;
            minCost = Math.min(minCost, calculator(y+1, idx));
        }
        accCostMap[y][x] =  minCost + costMap[y][x];
        return accCostMap[y][x];
    }


}

File: Kotlin/Silver/1149.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import kotlin.math.min


val br = BufferedReader(InputStreamReader(System.`in`))
lateinit var costMap: Array<IntArray>
lateinit var accCostMap: Array<IntArray>
var houseCount = 0

fun main() {

    houseCount = br.readLine().toInt()
    costMap = Array(houseCount) { IntArray(3) }
    for (idx in 0..houseCount-1) {
        costMap[idx] = br.readLine().split(" ").map { it.toInt() }.toIntArray()
    }

    accCostMap = Array(houseCount) { IntArray(3) }
    println(calculateCost())
    br.close()
}

fun calculateCost(): Int {
    var minCost = 987654321
    for (color in 0..2) {
        minCost = min(minCost, calculator(0, color))
    }
    return minCost
}

fun calculator(y: Int, x: Int): Int {

    if (y == houseCount) return 0
    if (accCostMap[y][x] != 0) return accCostMap[y][x]

    var minCost = 987654321
    for (color in 0..2){
        if (color == x) continue
        minCost = min(minCost, calculator(y+1, color))
    }
    accCostMap[y][x] = costMap[y][x] + minCost
    return accCostMap[y][x]
}